File: kotlinTutorials/kt10/src/main/java/com/java/kt10/LazyThreadSafeDoubleCheck.java
package com.java.kt10;

/**
 * 懒加载+线程安全+双重检查机制
 */
public class LazyThreadSafeDoubleCheck {
    private static volatile LazyThreadSafeDoubleCheck instance;
    private LazyThreadSafeDoubleCheck(){}

    public static LazyThreadSafeDoubleCheck getInstance() {
        if (null == instance) {
            synchronized (LazyThreadSafeDoubleCheck.class) {
                if (null == instance) {
                    instance = new LazyThreadSafeDoubleCheck();
                }
            }
        }
        return instance;
    }
}

File: kotlinTutorials/kt10/src/main/kotlin/com/kotlin/kt10/LazyThreadSafeDoubleCheck.kt
package com.kotlin.kt10

class LazyThreadSafeDoubleCheck private constructor(){

    companion object {
        val instance by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            LazyThreadSafeDoubleCheck()
        }
        
        @Volatile
        private var instance2 : LazyThreadSafeDoubleCheck ? = null

        fun getInstance2() : LazyThreadSafeDoubleCheck {
            if (null == instance2) {
                synchronized(this) {
                    if (null == instance2) {
                        instance2 = LazyThreadSafeDoubleCheck()
                    }
                }
            }
            return instance2!!
        }
    }
}